docs/src/proposals/return-data.md

```diff
@@ -0,0 +1,139 @@
+around through CPI. Again this does not work for CPI as the callee may not have
+permission to write to it.
+
+The problem with this solution is:
+
+- It does not work for RPC calls
+- It is very racey; a client has to submit the Tx and then retrieve the account
+  data. This is not atomic so the return data can be overwritten by another transaction.
+
+## Requirements for Solution
+
+It must work for:
+
+- RPC: An RPC should be able to return any number of values without writing to account data
+- Transaction: An transaction should be able to return any number of values without needing to write them account data
+- CPI: The callee must "set" return value, and the caller must be able to retrieve it.
+
+## Review of other chains
+
+### Ethereum (EVM)
+
+The `RETURN` opcode allows a contract to set a buffer as a returndata. This opcode takes a pointer to memory and a size. The `REVERT` opcode works similarly but signals that the call failed, and all account data changes must be reverted.
+
+For CPI, the caller can retrieve the returned data of the callee using the `RETURNDATASIZE` opcode which returns the length, and the `RETURNDATACOPY` opcode, which takes a memory destination pointer, offset into the returndata, and a length argument.
+
+Ethereum stores the returndata in blocks.
+
+### Parity Substrate
+
+The return data can be set using the `seal_return(u32 flags, u32 pointer, u32 size)` syscall.
+- Flags can be 1 for revert, 0 for success (nothing else defined)
+- Function does not return
+
+CPI: The `seal_call()` syscall takes pointer to buffer and pointer to buffer size where return data goes
+ - There is a 32KB limit for return data.
+
+Parity Substrate does not write the return data to blocks.
+
+## Rejected Solution
+
+The concept of ephemeral accounts has been proposed a solution for this. This would
+certainly work for the CPI case, but this would not work RPC or Transaction case.
+
+## Proposed Solution
+
+The callee can set the return data using a new system call `sol_set_return_data(u8 *buf, u64 length)`.
+This function does not return and signals success for execution of the contract. There is a limit
+of 1024 bytes for the returndata.
+
+The caller can retrieve the return data of the callee by using a new variant of the `sol_invoke` system call:
+
+```
+fn sol_invoke_signed_return_data_rust(
+    instruction_addr: *const u8,
+    account_infos_addr: *const u8,
+    account_infos_len: u64,
+    signers_seeds_addr: *const u8,
+    signers_seeds_len: u64,
+    return_data_addr: *mut u8,
+    return_data_length: *mut u64,
+) -> u64;
+
+uint64_t sol_invoke_signed_return_data_c(
+  const SolInstruction *instruction,
+  const SolAccountInfo *account_infos,
+  int account_infos_len,
+  const SolSignerSeeds *signers_seeds,
+  int signers_seeds_len
+  uint8_t *return_data,
+  uint64_t *return_data_length,
+);
+```
+On entry, `return_data_length` should point to the size of the buffer at `return_data`. If the callee
```

Comment:
Does the size of the return data impact compute unit consumption?

Reply:
That's an interesting question but I don't know what the answer to that is. @jackcmay @mvines wdyt?

Reply:
Rather than overloading invoke, why not break this out into a separate syscall that retrieves the return data?

Reply:
Overloading invoke saves the program making two syscalls (invoke & get_return_data) rather than just invoke.

Maybe the separate syscall makes more sense. Happy to move on that and amend the proposal.

Reply:
Yeah, that's the trade-off.  I'm leaning toward separating the two so that the two API's don't get entwined.  Especially if we change invoke, we might need to create two new invokes.  But the advantage of a single syscall isn't negligible so I could be swayed either way.

Reply:
And yes, we have a common way of charging for data passed in and out of the program, these bytes should incur a similar cost

Reply:
I've updated the proposal.